src/main/java/dk/cngroup/trainings/spring/springassignment/controller/AnimalController.java

```diff
@@ -14,61 +17,102 @@
-        if(searchedAnimal!=null)
-            return new ResponseEntity<>(searchedAnimal.get(),HttpStatus.OK);
-        else
-            return ResponseEntity.notFound().build();
-    }
+	@RequestMapping(path = "/{id}", method = RequestMethod.GET)
+	public ResponseEntity<Animal> getAnimalById(@PathVariable("id") long id) {
+		Optional<Animal> searchedAnimal = animalService.getAnimalById(id);
+		if (searchedAnimal.isPresent()) {
+			return new ResponseEntity<>(searchedAnimal.get(), HttpStatus.OK);
+		} else {
+			return ResponseEntity.notFound().build();
+		}
+	}
 
-    @RequestMapping(path = "/search/{name}", method = RequestMethod.GET)
-    public ResponseEntity<List<Animal>> getAnimalByName(@PathVariable("name") String name){
-        List<Animal> sameNameAnimals= animalService.getAnimalsByName(name);
-        if(sameNameAnimals.size()!=0)
-            return new ResponseEntity<>(sameNameAnimals, HttpStatus.OK);
-        else
-            return ResponseEntity.notFound().build();
-    }
+	@RequestMapping(path = "/name", method = RequestMethod.GET)
+	public ResponseEntity<List<Animal>> getAnimalsByName(@RequestParam("name") String name) {
+		List<Animal> sameNameAnimals = animalService.getAnimalsByName(name);
+		if (!sameNameAnimals.isEmpty()) {
+			return new ResponseEntity<>(sameNameAnimals, HttpStatus.OK);
+		} else {
+			return ResponseEntity.notFound().build();
+		}
+	}
+
+	@RequestMapping(path = "/{id}/careTakers", method = RequestMethod.POST)
+	public ResponseEntity<CareTaker> addNewCareTakerToExistingAnimal(@PathVariable("id") long id,
+																	 @RequestBody @Valid CareTaker careTaker) {
+		if (animalService.getAnimalById(id).isPresent()) {
+			Optional<CareTaker> addedCareTaker = null;
+			try {
+				addedCareTaker = animalService.addNewCareTakerToExistingAnimal(id, careTaker);
+				return new ResponseEntity<>(addedCareTaker.get(), HttpStatus.OK);
+			} catch (InvalidCareTakerException e) {
+				return ResponseEntity.badRequest().build();
+			}
+		} else {
+			return ResponseEntity.notFound().build();
+		}
+	}
+
+	@RequestMapping(value = "/{id}/careTakers", method = RequestMethod.GET)
+	public ResponseEntity<List<CareTaker>> getCareTakersByAnimalId(@PathVariable("id") long id) {
+		Optional<Animal> animalById = animalService.getAnimalById(id);
+		if (animalById.isPresent()) {
+			return new ResponseEntity<>(animalById.get().getCareTakers()
+					, HttpStatus.OK);
+		} else {
+			return ResponseEntity.notFound().build();
+		}
+	}
+
+	@RequestMapping(path = "/{animalId}/careTakers/{careTakerId}", method = RequestMethod.PUT)
+	public ResponseEntity<String> addExistingCareTakerToExistingAnimal(@PathVariable("animalId") long animalId,
+																	   @PathVariable("careTakerId") long careTakerId) {
+		System.out.println(">>>>>>> From animalController");
+		String result = animalService.addExistingCareTakerToExistingAnimal(animalId, careTakerId);
+		if (result.startsWith("Success")) {
```

Comment:
I think that in the last review we decided, that matching on String is bad. Use an exception instead.

Reply:
Fixed.